### PR TITLE
Show OTP delay notification

### LINKED ISSUE #629: Detect possible OTP delay and show message on login page
thinking this is something @jia1 can help to implement

thoughts on cheapest way to detect? @jia1 @karrui 

**Updated**

- **25s** after OTP requested, show blue infobox with copy below
- every time we show this notification, log timestamp and email address so we have a gauge of OTP delay occurrences
- this otp delay infobox disappears if OTP is sent again and the timer resets
- on login page, change link text 'Have a question?' to 'Need help?'

Copy:
'OTP might be delayed due to government email traffic. Please try again later, or contact us if issue persists.'

### PR DESCRIPTION
**Pull request by jia1**
## Problem

Closes #629

## Solution

At the finally clause of the send OTP function, add a timeout function which sets the show OTP delay notification flag to true after 20 seconds.

Priority of `vm.signInMsg.isMsg`, `vm.showOtpDelayNotification` and `vm.isOtpSending` is described in the conversation below ([link](https://github.com/opengovsg/FormSG/pull/748#discussion_r537451785)).

## Before & After Screenshots

**AFTER**:

https://user-images.githubusercontent.com/14961285/103256818-84c9bd80-49c9-11eb-9a4e-afa63584e3ac.mov

### CODE REVIEW

**Comment by syan-syan** [comment]
sorry can we add 'contact us if `the` issue persists' 

thanks!

File: src/public/modules/users/views/authentication/signin.client.view.html [hunk 1]
```diff
@@ -82,7 +82,7 @@
           ng-keyup="$event.keyCode == 13 && vm.credentials.otp.length === 6 && vm.verifyOtp()"
         />
         <div
-          ng-if="vm.signInMsg.isMsg"
+          ng-if="!vm.showOtpDelayNotification && vm.signInMsg.isMsg"
```
Comment (karrui) [inline src/public/modules/users/views/authentication/signin.client.view.html:85]: i'd think these messages take precedence over the delay notification. Maybe reverse the conditional? only show otpDelayNotification if !vm.isOtpSending || !vm.signInMsg.isMsg

Comment (jia1) [reply src/public/modules/users/views/authentication/signin.client.view.html]: I made adjustments such that the `ng-if`s should behave like the following table:

| vm.signInMsg.isMsg | vm.showOtpDelayNotification | vm.isOtpSending |                        |
|--------------------|-----------------------------|-----------------|------------------------|
| 0                  | 0                           | 0               | Nothing                |
| 0                  | 0                           | 1               | Sending OTP...         |
| 0                  | 1                           | 0               | OTP delay notification |
| 0                  | 1                           | 1               | OTP delay notification |
| 1                  | 0                           | 0               | Sign-in messge         |
| 1                  | 0                           | 1               | Sign-in message        |
| 1                  | 1                           | 0               | Sign-in message        |
| 1                  | 1                           | 1               | Sign-in message        |

I kind of did this up in a slight rush so I will be back to verify my work once more.

**Comment by jia1** [comment]
> sorry can we amend copy to
> 
> OTP might be delayed due to government email traffic. Please try again later, or contact us `if you're unable to log in. `
> 
> thanks!

It has been done. I will add the screenshot later.

File: src/public/modules/users/views/authentication/signin.client.view.html (hunk 1 above)
Comment (karrui) [reply src/public/modules/users/views/authentication/signin.client.view.html]: I see. I think will need more design thinking to see which message has priority. Maybe if both sending messages and showOtpDelayNotification will need a combined message. @syan-syan thoughts?

Maybe just concat `OTP might be delayed due to government email traffic` to all current messages after 45 seconds.

e.g. sending otp after message:
` Sending OTP... OTP might be delayed due to government email traffic` etc etc

**Comment by karrui** [comment]
Build fails due to stuff on develop branch, will fix this branch (and new lint rules due to dependency updates)

**Comment by pearlyong** [comment]
@jia1 can we reduce the time it takes to show the info box from 45 seconds to 20 seconds? 
20 seconds seems like a reasonable time for a user to wait without them clicking on Resend OTP.
thanks!

**Comment by jia1** [comment]
@pearlyong @syan-syan @karrui I have updated the delay from 45s to 20s and uploaded a screen recording.

**Review by karrui — APPROVED** [review]
nice! lgtm

**Comment by syan-syan** [comment]
hey @jia1 can we make two quick amendments
- remove exclamation mark behind the message "OTP sent to email@email.com!" so it looks like "OTP sent to email@email.com"?
- amend OTP delay message to "OTP might be delayed due to government email traffic. If you are unable to sign in, please try again later, or contact us."

File: src/public/modules/users/controllers/authentication.client.controller.js [hunk 2]
```diff
@@ -161,40 +162,44 @@ function AuthenticationController($state, $timeout, $window, Auth, GTag) {
...
+            'Failed to send login OTP. Please try again later and if the problem persists, contact us.'
+          vm.signInMsg = {
+            isMsg: true,
+            isError: true,
+            msg,
+          }
+        },
+      )
+      .finally(function () {
         $timeout(function () {
```
Comment (karrui) [inline src/public/modules/users/controllers/authentication.client.controller.js:198]: Sorry, can you assign this timeout to a variable so we can cancel it when the login succeeds

```ts
// Declared variable outside: `let delayNoticePromise`

delayNoticePromise = $timeout(function () {
  ...
})
```

File: src/public/modules/users/controllers/authentication.client.controller.js [hunk 3]
```diff
@@ -161,40 +162,44 @@ function AuthenticationController($state, $timeout, $window, Auth, GTag) {
...
-        }, 100)
-      },
-    )
+          vm.signInMsg.isMsg = false
+          vm.showOtpDelayNotification = true
+        }, 20000)
+      })
   }
 
   /**
```
Comment (karrui) [inline src/public/modules/users/controllers/authentication.client.controller.js:205]: Then when the user logs in successfully/page navigates away

```
$scope.$on('$destroy', function() {
    $timeout.cancel(delayNoticePromise);
});
```

**Comment by karrui** [comment]
I see you merged it, haha! No worries, might do a second PR for that

**Pull request merged**